services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java

```diff
@@ -360,11 +360,45 @@ public BrokeredIdentityContext getFederatedIdentity(String response) {
             throw new IdentityBrokerException("Could not decode access token response.", e);
         }
         String accessToken = verifyAccessToken(tokenResponse);
-
         String encodedIdToken = tokenResponse.getIdToken();
 
         JsonWebToken idToken = validateToken(encodedIdToken);
 
+        // We need to manually retrieve groups of the user.
+        // See https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens#groups-overage-claim
```

Comment:
This logic should be added to the `MicrosoftIdentityProvider` provider.

Reply:
Done.

Reply:
Moving the feature completely to MS Identity provider would not solve our problem, because we are using the "plain" oidc provider to connect to azure. Maybe it would be possible to use the code in booth MS and "plain" oidc identity provider by implementing a filter in "plain" oidc provider like `if issuer contains "microsoft" then <trigger code>` to preserve it there.

Guess there were reasons why we choose not to use the ms provider, I'll try to figure out why and get back with more information.

Reply:
When configuring our Keycloak instance initially, my colleagues were of the opinion that the MS provider is only for usage with private MS account, as pointed out here https://keycloak.discourse.group/t/is-microsoft-identity-provider-using-openid-connect-v1-0/21933/3. 

This was not what we needed, so we choose the generic oidc provider.

Reply:
I see. What MS broker is missing to support your case then? Sorry, if I'm missing something.

If it is about making it more configurable in terms of which URLs should be used to certain OIDC/OAuth2 endpoints (e.g.: IIRC that discussion), we should probably consider improving it.

But yeah, it does not make much sense to have code related to a specific OP within the generic broker ... People here won't accept this.

Reply:
I had a discussion with a colleague about this. He also said the the MS Identity Provider is really for social login only. For Azure AD/Entra ID, you might need multiple providers per realm and target different tenants in Entra ID for example. We use our own AzureADIdentityProvider which is working fine and also supports the overage indicator. Would it be an idea to just add a new IdentityProvider for this? There are other specifics on Entra ID like using logout_hint instead if id_token_hint on logout. I think we should be able to just contribute the complete implementation we already have.

Reply:
Are there more opinions on this? Can I close this pr because it won't get accepted anyway?

Reply:
@pedroigor What do you think about adding an additional IDP for Azure ID/Entra ID? We have a full implementation and if I remember correctly @thomasdarimont also has one.

Reply:
@NilsGriebner @pedroigor 
I think it would be a greate contribution to improve the OpenID Connect provider and add `Aggregated and Distributed Claims`, because their usage is not specific to Entra.
https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims

Entra uses the feature correctly, but because of deprecation of `Azure AD Graph` and not recommended `implicit flow`, there has been some hacks around, thus your need to make a dedicated Provider.
https://learn.microsoft.com/en-us/security/zero-trust/develop/configure-tokens-group-claims-app-roles#group-overages

My suggestion would be to:
- Improve OpenID Connect provider with  `Aggregated and Distributed Claims`
- Add a *temporary* Dedicated Entra Provider that will use OIDC provider and make the call `Microsoft Graph` to get groups.
- *Wait* for Azure to fully remove the deprecated `Azure AD Graph` and provides the correct url in `_claim_sources`.


But still
> If you're using OpenID Connect or OAuth2, you can have up to 200 groups in your token. If you're using SAML, you can have only 150 groups because SAML tokens are bigger than OAuth2 and OpenID Connect tokens. If you're using the implicit flow, the limit is six because those responses show up in the URL. In all of these cases, instead of having a groups claim, you'll see an indication (known as a group overage) that tells you that the user is a member of too many groups to fit in your token.

I am curious of your usage to have 200+ groups, and need to map them in keycloak.

For my usage, i just assign only needed groups to the application to avoid unnecessary claims in the token.

> One way to avoid hitting a group overage claim is to select Groups assigned to the application on the Edit groups claim screen instead of All groups.

Reply:
@sschu Thanks for the heads up :) I did in fact create a custom OIDC IdpMapper for resolving the group information provided by EntraID via the msgraph API:
https://github.com/thomasdarimont/keycloak-project-example/blob/main/keycloak/extensions/src/main/java/com/github/thomasdarimont/keycloak/custom/idp/azure/CustomAzureADGroupMapper.java

For most of my AAD/EntraID integrations, I also use the plain OIDC IdP support, sometimes in combination with my custom group mapper.

Reply:
I usually don't expose a large number of groups in the IDToken, as I consider this to be bad pratice.
In such cases I'd rather store the AAD refresh token in the provider link and create a custom (OIDC) protocol mapper, which can then use the AAD token of the user to dynamically fetch group information via the graph API. The protocol mapper is then configured to only expose the group information via the user-info endpoint.

Reply:
> I am curious of your usage to have 200+ groups, and need to map them in keycloak.

We don't have access to all Entra Id accounts connecting to our Keycloak and therefore can't incfluence how groups are assigned.

Reply:
Sorry for the delay. By reading the discussions, I understand that:

* We are missing https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims
* Even though we support 1) above it is not enough to support AAD/Entra
* AAD/Entra have issues and need workarounds to properly handle 1) the above

If all that is true, I would vote for leveraging the existing implementations from @sschu and @thomasdarimont. 
@NilsGriebner Regardless of the approach, providing a test for this will be tough as we have issues running broker tests when they depend on too many configurations on the IdP side.

Wdyt?

Reply:
+1 for leveraging  existing implementations.